pkg/storage/s3.go

```diff
@@ -1,56 +1,288 @@
-	flags.String(s3EndpointOption, "", "Set the AWS S3 endpoint URL")
-	flags.String(s3RegionOption, "", "Set the AWS region")
-	// TODO: Finalize the list of options.
-	_ = flags.MarkHidden(s3EndpointOption)
+	flags.String(s3EndpointOption, "", "Set the S3 endpoint URL, please specify the http or https scheme explicitly")
+	flags.String(s3RegionOption, "", "Set the S3 region, e.g. us-east-1")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class, e.g. STANDARD")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm, e.g. AES256")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs, e.g. authenticated-read")
+	flags.String(s3ProviderOption, "", "Set the S3 provider, e.g. aws, alibaba, ceph")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	sendCredential, err = flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
 	options.Endpoint, err = flags.GetString(s3EndpointOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
 	options.Region, err = flags.GetString(s3RegionOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
+	options.SSE, err = flags.GetString(s3SSEOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ACL, err = flags.GetString(s3ACLOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.StorageClass, err = flags.GetString(s3StorageClassOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	options.ForcePathStyle = true
+	options.Provider, err = flags.GetString(s3ProviderOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
 
-	// TODO: Add more options here.
+	return options, err
+}
 
-	return
+// newS3Storage initialize a new s3 storage for metadata
+func newS3Storage(backend *backup.S3) (*S3Storage, error) {
+	qs := *backend
+	var cred *credentials.Credentials
+	if qs.AccessKey != "" && qs.SecretAccessKey != "" {
+		cred = credentials.NewStaticCredentials(qs.AccessKey, qs.SecretAccessKey, "")
+	}
+	awsConfig := aws.NewConfig().
+		WithMaxRetries(maxRetries).
+		WithS3ForcePathStyle(qs.ForcePathStyle).
+		WithRegion(qs.Region)
+	if qs.Endpoint != "" {
+		awsConfig.WithEndpoint(qs.Endpoint)
+	}
+	if cred != nil {
+		awsConfig.WithCredentials(cred)
+	}
+	// awsConfig.WithLogLevel(aws.LogDebugWithSigning)
+	awsSessionOpts := session.Options{
+		Config: *awsConfig,
+	}
+	ses, err := session.NewSessionWithOptions(awsSessionOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	if sendCredential && ses.Config.Credentials != nil {
+		if qs.AccessKey == "" || qs.SecretAccessKey == "" {
+			v, cerr := ses.Config.Credentials.Get()
+			if cerr != nil {
+				return nil, cerr
+			}
+			backend.AccessKey = v.AccessKeyID
+			backend.SecretAccessKey = v.SecretAccessKey
+		}
+	}
+
+	c := s3.New(ses)
+	err = checkS3Bucket(c, qs.Bucket)
+	if err != nil {
+		return nil, errors.Errorf("checkS3Bucket error: %v", err)
```

Comment:
```suggestion
		return nil, errors.Errorf("Bucket %s is not accessible: %v", qs.Bucket, err)
```

The error message is visible to the CLI users.